File: scala/ex3_4.scala
trait Timestamper:
  type T
  def init: T
  def next(t: T): T
  def cmp(t1: T, t2: T): Boolean

class IntTimestamper extends Timestamper:
  type T = Int
  def init: T = 0
  def next(t: T): T = t + 1
  def cmp(t1: T, t2: T): Boolean = t1 < t2

class StringTimestamper extends Timestamper:
  type T = String
  def init: T = "a"
  def next(t: T): T = t + "a"
  def cmp(t1: T, t2: T): Boolean = t1.length < t2.length

def create(): Timestamper =
  if true then IntTimestamper() else StringTimestamper()

@main def main =
  val t: Timestamper = create()
  val ts0: t.T = t.init
  val ts1: t.T = t.next(ts0)
  t.cmp(ts0, ts1)

File: ocaml/ex3_4.ml
module type TIMESTAMPER = sig
  type t
  val init: t
  val next: t -> t
  val cmp: t -> t -> bool
end

module IntTimestamper = struct
  type t = int
  let init = 0
  let next t = t + 1
  let cmp t1 t2 = t1 < t2
end

module StringTimestamper = struct
  type t = string
  let init = "a"
  let next t = String.cat t "a"
  let cmp t1 t2 = (String.length t1) < (String.length t2)
end

let create (_: unit): (module TIMESTAMPER) =
  if true then (module IntTimestamper) else (module StringTimestamper)

let _ =
  let module T = (val create (): TIMESTAMPER) in
  let ts0: T.t = T.init in
  let ts1: T.t = T.next ts0 in
  T.cmp ts0 ts1
